yellow-paper/docs/gas-and-fees/gas-metering-and-fees.md

```diff
@@ -0,0 +1,409 @@
+```
+
+
+It will be the same for the `compute_gas_used`.
+
+### Out of gas
+
+Public execution must stop as soon as possible once `daGasLeft` is exhausted.
+
+The public VM will revert if an exception is thrown, including an out of gas exception.
+When this occurs, the public VM will return a revert code.
+
+If this occurs in setup, the transaction is invalid and cannot be included in the block.
+
+If this occurs in app_logic, `daGasLeft` is set to 0, and thus `public_kernel_public_inputs.gas_limits.da_gas_limit == public_kernel_public_inputs.end.gas_used.da_gas_used + public_kernel_public_inputs.end_non_revertible.gas_used.da_gas_used`. 
+
+Similarly for `computeGasLeft`.
+
+In the teardown phase, there cannot be an out of gas error because the teardown phase effectively has infinite gas (since the user has prepaid for it via `inclusion_fee`). If there is any other exception though, the transaction is invalid and cannot be included in the block. It is thus expected that sequencers will maintain a whitelist of teardown functions and the minimum `inclusion_fee` required for each.
+
+### Recombining for teardown
+
+The public kernel **tail** will recombine the `gas_used` values from the non-revertible and revertible phases.
+
+```
+kernel_public_inputs.gas_limits := public_kernel_public_inputs.gas_limits
+
+kernel_public_inputs.end.gas_used.da_gas_used =
+    public_kernel_public_inputs.end_non_revertible.gas_used.da_gas_used
+    + public_kernel_public_inputs.end.gas_used.da_gas_used
+
+kernel_public_inputs.end.gas_used.compute_gas_used :=
+    public_kernel_public_inputs.end_non_revertible.gas_used.compute_gas_used +
+    + public_kernel_public_inputs.end.gas_used.compute_gas_used
+
+
+fee_per_da_gas := public_kernel_public_inputs.constants.historical_header.global_variables.fee_per_da_gas
+fee_per_compute_gas := public_kernel_public_inputs.constants.historical_header.global_variables.fee_per_compute_gas
+
+kernel_public_inputs.transaction_fee :=
+    fee_per_da_gas * kernel_public_inputs.end.gas_used.da_gas_used
+    + fee_per_compute_gas * kernel_public_inputs.end.gas_used.compute_gas_used
+    + kernel_public_inputs.gas_limits.inclusion_fee
+```
+
+# Teardown
+
+After executing the teardown function, the teardown kernel will assert that the `transaction_fee` that was made available in the public context is equal to `kernel_public_inputs.transaction_fee`.
+
+This ensures that the `pay_fee` function on the enshrined gas token contract **can** be called with the correct amount.
+
+### Validating fee payment
+
+During execution of the teardown function, the public VM will need to inform the kernel **if** the `pay_fee` function was called.
+
+This is done via a `feePaid` boolean in the MachineState.
+
+This field is included in the `PublicCircuitPublicInputs` struct as `fee_paid`.
+
+The public kernel teardown will then set:
+
+```
+kernel_public_inputs.transaction_fee :=
+    public_inputs.fee_paid ? kernel_public_inputs.transaction_fee : 0
```

Comment:
I find it confusing that the field `transaction_fee` can mean "how much you have to pay" or "how much have you paid", depending on where you see it.

Reply:
Also: can't we drop the `feePaid` flag and just zero out the `transaction_fee_paid` to convey that the fee wasn't paid?